src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -0,0 +1,176 @@
+import logging
+from datetime import datetime, timedelta
+from typing import Callable, Mapping, MutableMapping, Optional, Sequence, TypedDict, Union, cast
+
+from arroyo import Topic
+from arroyo.backends.kafka import KafkaConsumer, KafkaPayload
+from arroyo.backends.kafka.configuration import build_kafka_consumer_configuration
+from arroyo.processing import StreamProcessor
+from arroyo.processing.strategies import ProcessingStrategy, ProcessingStrategyFactory
+from arroyo.types import Message, Partition, Position
+from django.conf import settings
+
+from sentry.constants import DataCategory
+from sentry.sentry_metrics.indexer.strings import TRANSACTION_METRICS_NAMES
+from sentry.utils import json
+from sentry.utils.outcomes import Outcome, track_outcome
+
+logger = logging.getLogger(__name__)
+
+
+def get_metrics_billing_consumer(
+    group_id: str,
+    auto_offset_reset: str,
+    force_topic: Union[str, None],
+    force_cluster: Union[str, None],
+    max_batch_size: int,
+    max_batch_time: int,
+) -> StreamProcessor[KafkaPayload]:
+    topic = force_topic or settings.KAFKA_SNUBA_GENERIC_METRICS
+    bootstrap_servers = _get_bootstrap_servers(topic, force_cluster)
+
+    return StreamProcessor(
+        consumer=KafkaConsumer(
+            build_kafka_consumer_configuration(
+                default_config={},
+                group_id=group_id,
+                auto_offset_reset=auto_offset_reset,
+                bootstrap_servers=bootstrap_servers,
+            ),
+        ),
+        topic=Topic(topic),
+        processor_factory=BillingMetricsConsumerStrategyFactory(max_batch_size, max_batch_time),
+    )
+
+
+def _get_bootstrap_servers(topic: str, force_cluster: Union[str, None]) -> Sequence[str]:
+    cluster = force_cluster or settings.KAFKA_TOPICS[topic]["cluster"]
+
+    options = settings.KAFKA_CLUSTERS[cluster]
+    servers = options["common"]["bootstrap.servers"]
+    if isinstance(servers, (list, tuple)):
+        return servers
+    return [servers]
+
+
+class BillingMetricsConsumerStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
+    def __init__(self, max_batch_size: int, max_batch_time: int):
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+
+    def create_with_partitions(
+        self,
+        commit: Callable[[Mapping[Partition, Position]], None],
+        partitions: Mapping[Partition, int],
+    ) -> ProcessingStrategy[KafkaPayload]:
+        return BillingTxCountMetricConsumerStrategy(
+            commit, self.__max_batch_size, self.__max_batch_time
+        )
+
+
+class MetricsBucket(TypedDict):
+    """
+    Metrics bucket as decoded from kafka.
+
+    Only defines the fields that are relevant for this consumer."""
+
+    org_id: int
+    project_id: int
+    metric_id: int
+    timestamp: int
+    value: Sequence[float]  # only true for distribution buckets
```

Comment:
Since this isn't true for all buckets observed, could we leave that untyped and introduce an explicit check, or does that increase overheads too much?

Reply:
Done.